src/mtype.c

```diff
@@ -3980,6 +3980,8 @@ void TypeSArray::resolve(Loc loc, Scope *sc, Expression **pe, Type **pt, Dsymbol
     }
     else
     {
+        if ((*pt)->ty != Terror)
+            next = *pt;     // prevent re-running semantic() on 'next'
```

Comment:
Indeed this change looks hacky - but this change follows current compiler design. 

Currently `(Expression|Type|Dsymbol)::semantic()` step will cause two side effects at least - the deprecation and information warning message. They don't affect the code meaning. Then, if semantic() step is invoked twice on same AST entity, they would make duplicated console outputs. To avoid the issue, we should cache the valid semantic result in each steps. After that, the repeated semantic analysis on the cached AST entity will be stopped by the previous flagging, that is (setting `Expression::type` to non-null | merge `Type` objects and set `Type::deco` to non-null | the increment of `Dsymbol::semanticRun` value).

In here, if the semantic result in `*pt` is valid type, it is cached into `this->next` to prevent rerunning semantic and messaging side effects. From that reason, this change is clean fix to me.

And, it's free that you think the better dmd internal design. Most of my bugfix work in several years was to fill holes in the well designed but rough implemented compiler internals. Honestly, I cannot say I'm genious architect programmer to design enough clean & efficient compiler process. Better design is welcome.